### PR TITLE
Implement Serializable trait for all structs

### LINKED ISSUE #56: Implement `to` / `from_bytes`from dusk-bytes
Every sized structure in JubJub that implement manually `from_bytes` and `to_bytes` should use the traits exposed by `dusk-bytes` instead.

That also means get rid of `CtOption` for the `from_bytes` result.

### PR DESCRIPTION
**Pull request by julesdesmit**
Fixes #56

### CODE REVIEW

File: src/elgamal.rs [hunk 1]
```diff
@@ -69,6 +70,52 @@ pub struct ElgamalCipher {
...
+
+    /// Deserialize from a [`ElgamalCipher::to_bytes`] construction
+    fn from_bytes(bytes: &[u8; Self::SIZE]) -> Result<Self, Self::Error> {
+        let mut gamma = [0u8; 32];
+        let mut delta = [0u8; 32];
+
+        gamma.copy_from_slice(&bytes[..32]);
+        delta.copy_from_slice(&bytes[32..]);
+
+        let gamma = JubJubAffine::from_bytes(&gamma);
```
Comment (vlopes11) [inline src/elgamal.rs:100]: Maybe do a `JubJubAffine::from_slice(&bytes[..32])` instead of executing a new copy?

Comment (vlopes11) [inline src/elgamal.rs:100]: Consider transforming the `CtOption` into `Option` so you can avoid unwraping.

`let gamma: Option = JubJubAffine::from_slice(&bytes[..32]).into()`

**Review by vlopes11 — APPROVED** [review]
LGTM, few nits

File: src/elgamal.rs (hunk 1 above)
Comment (julesdesmit) [reply src/elgamal.rs:100]: Wouldn't I still have to unwrap, since it is an `Option`?

Comment (julesdesmit) [reply src/elgamal.rs:100]: I can't use `from_slice` in the `Serializable` implementation, because there's a cyclic dependency.

File: RELEASES.md [hunk 2]
```diff
@@ -1,3 +1,11 @@
+# 0.7.0
+### Change
+- Update `dusk-bls12_381` to `0.5.0`.
+- Downgrade `rand` to `0.7`.
+- Downgrade `rand_core` to `0.5`.
+- Downgrade `rand_xorshift` to `0.2`.
+- Closes #56, implements `Serializable` for all structures.
```
Comment (ZER0) [inline RELEASES.md:7]: I would avoid to mentioned issues like that, they don't get linked to the actual issue. Also, they should just mentioned at the end. I noticed it was doing in the past but it's not correct.

I'll be sure to formalize the changelog format once we finish with priorities, but the important thing here is that you're doing a breaking change in the API – `from_bytes` DO NOT return a `Option` / `CtOption` anymore. That's the important thing to highlight.
To me this is translating in three different entries in the changelog, something like:

```
### Added
- Add `Serializable` trait to all structures

### Removed
- Remove manual implementation of `from_bytes` and `to_bytes` from all structures

### Changed
- Change return value of `from_bytes` from  `Option` / `CtOption` into `Result<Self, Error>`
```

Something similar to that.

Notice also that where usually the verb of action itself is present ("Change", "Add", etc) the section is using the past tense ("Changed", "Added", etc).

But this is more about the convention, we can postpone the refactor of the changelog for later. But I would rephrase the change in a better way.

If you want a generic reference, you can use: https://keepachangelog.com/en/1.0.0/

File: src/lib.rs [hunk 3]
```diff
@@ -553,8 +526,47 @@ impl JubJubAffine {
...
                 CtOption::new(JubJubAffine { x: final_x, y }, Choice::from(1u8))
-            })
-        })
+            }),
+        )
+        .ok_or(BytesError::InvalidData)
+    }
+}
+
+impl DeserializableSlice<32> for JubJubAffine {}
```
Comment (ZER0) [inline src/lib.rs:535]: Remove this, is not necessary – it would actually result in a compilation error now.

Use `cargo update` locally to update your deps, so you can see that it will fails.

File: src/fr.rs [hunk 4]
```diff
@@ -62,7 +63,7 @@ impl From<Fr> for BlsScalar {
         // so convert any jubjub scalar to a BLS' Scalar should always be
         // safe.
         assert!(
-            bool::from(bls_scalar.is_some()),
+            bool::from(bls_scalar.is_ok()),
```
Comment (ZER0) [inline src/fr.rs:66]: `is_ok` returns already a boolean, this is an unnecessary conversion, just have `assert!(bls_scalar.is_ok(), "Failed to convert a Scalar from JubJub to BLS");`

File: src/lib.rs [hunk 5]
```diff
@@ -1386,14 +1398,16 @@ fn second_gen_nums() {
         hasher.update(counter.to_le_bytes());
         let res = hasher.finalize();
         array.copy_from_slice(&res[0..32]);
-        if JubJubAffine::from_bytes(array.clone()).is_some().into()
-            && JubJubAffine::from_bytes(array)
+        if JubJubAffine::from_bytes(&array).is_ok().into()
```
Comment (ZER0) [inline src/lib.rs:1401]: `is_ok()` is already a boolean, `.into()` is unnecessary.

**Review by ZER0 — APPROVED** [review]
LGTM but there are few changes to do.

**Review by ZER0 — CHANGES_REQUESTED** [review]
Forgot to add "request changes" :)

File: src/lib.rs (hunk 3 above)
Comment (julesdesmit) [reply src/lib.rs:535]: Apparently I forgot to commit that 😅 thought I already did

File: RELEASES.md [hunk 6]
```diff
@@ -1,3 +1,13 @@
+# 0.7.0
+### Added
```
Comment (ZER0) [inline RELEASES.md:2]: Sorry my comment was a bit confusing probably: I meant that we should use this naming convention, but currently we're not using that. So it's better to keep consistent the naming and then refactoring the whole changelog (unless you want to refactor now). So what I'm suggesting is, use the present here as well as it was done in the other part of the same file; but we should in the future, when we have formalized in the way the changelog needs to be done, be more consistent with the format suggested in https://keepachangelog.com/en/1.0.0/.

So, my suggestion would be rename `Added` to `Add`, `Removed` to `Remove` and `Changed` to `Change`. Unless you want to rename also the past section as well – that would be fine too.
The important thing is the consistency.

**Review by ZER0 — APPROVED** [review]
